### DISCUSSION TITLE
Proper Masking Region

### DISCUSSION THREAD

**Discussion by Rajit13** [AstroBridge/BDLensing#24] [Q&A] [answered]
Are masks here seem okay or do I need to further increase the radius and position of a few? I masked two sources using a single circle, will it be a problem?
![mask](https://user-images.githubusercontent.com/81336262/231047160-1df5a3ae-f8bf-47f2-b6e9-8513887e239f.jpg)

**Comment by ajshajib** [comment]
The masked galaxies need to be fully covered. I can see that the light from the two large galaxies at the top (and the one peeking out of the left side) are still noticeable outside the masked region. Once you fit your model, these "leakages" will be prominent in the residual. It is thus very common to readjust the mask a few times while working on improving the model fit.

**Reply by Rajit13** [reply]
I'm not sure, should I also remove this arc? There are more sources that are close to the central galaxy which I'll mask now
![image](https://user-images.githubusercontent.com/81336262/231131786-dffa2771-ba8e-4b13-9c9d-c39dcc65b432.png)

**Reply by ajshajib** [reply]
No, the arcs and the lensed images contain all the information about the lens galaxy's mass, so they should never be masked. You might have two source components here: 1 creating the arc and one counter image blob at 3 o'clock, and the other source creating two image blobs at 1 and 7 o'clock. You can first try to mask 1 and 7 o'clock blobs to make it a simpler case, but eventually, you would want to unmask these to use their information in the lens model.

**Reply by Rajit13** [reply]
This is my current situation with improved masks. No idea how this inner red ring appeared out of the blue! But now I am sure there is something wrong with the Convergence and Magnification plot. Current reduced X^2=10.45

![improved111](https://user-images.githubusercontent.com/81336262/231241713-51366d85-f24e-4af0-af8a-e9842398b0f7.png)

**Reply by ajshajib** [reply]
It would be helpful if you link your code/notebook here. I suspect, your upper limit on `'beta'` is too high, it should not need to be larger than ~0.3.

**Reply by ajshajib** [reply]
Can you commit your notebook to your fork and provide a GitHub link to the notebook here?

**Reply by Rajit13** [reply]
[Here ](https://github.com/Rajit13/BDLensing/blob/32f4fea6b10ba41d300a507145f99f9cc3d1211d/DESIJ2158+0257/BDLensing_DESIJ2158+0257.ipynb)is the current code [which shows invalid due to error messages. [Collab](https://colab.research.google.com/drive/1O81IjE4i36S2CaRAWSrS0S4hzaIz_POu?usp=sharing) file where I am currently working.

**Reply by Rajit13** [reply]
The current reduced X^2 is 3. The link above is viewable. Please suggest any changes.

**Reply by ajshajib** [reply]
In your mask above, try removing the mask at 3 o'clock right beside the galaxy and adding a mask for the blob at 7 o'clock.

**Reply by Rajit13** [reply]
I tried the whole day tweaking with parameters like Beta/gamma./R_sersic, and theta_E but still could not go under X^2=6. But here is what you asked me to do; for this plot the X^2=13

[Here is the code](https://github.com/Rajit13/BDLensing/blob/32f4fea6b10ba41d300a507145f99f9cc3d1211d/DESIJ2158+0257/BDLensing_DESIJ2158+0257.ipynb) for when I got the best result so far X^2=3

![image](https://user-images.githubusercontent.com/81336262/231546126-c5ea4650-4d81-4c20-b4ae-a0b6822d7658.png)

**Reply by ajshajib** [reply]
This system is indeed a hard one to model. It is normal to try for a long time before a good model is found. The last one I modeled took me two weeks.

Here are a few observations. You should make one of these changes at a time to develop intuition on how they are affecting the lens model. There is a strong residual at the center of the lensing galaxy, so one Sersic function is not enough, you may add a second Sersic function for the lens galaxy, fix the `n_sersic`s of the two Sersic profiles to 1 and 4 and join their centroids. The source shapelets `beta` is still too large, contributing to lensed arcs at scales larger than they should, so you may try setting the upper limit to an even lower value. The lens 'gamma' value is atypically low, you should fix it to 2 **for now**, while you are tweaking other things.

**Reply by Rajit13** [reply]
I tried to add a second sersic model but now I am getting `IndexError: list index out of range` 

Here is the cell

`
# lens galaxy's mass model
fixed_lens = []
kwargs_lens_init = []
kwargs_lens_sigma = []
kwargs_lower_lens = []
kwargs_upper_lens = []

fixed_lens.append({})
kwargs_lens_init.append({'theta_E': 1.3, 'gamma': 2, 'e1': 0.01, 'e2': 0.,
                         'center_x': 0., 'center_y': 0.})
kwargs_lens_sigma.append({'theta_E': .2, 'gamma': .1, 'e1': 0.05, 'e2': 0.05,
                         'center_x': 0.5, 'center_y': 0.5})
kwargs_lower_lens.append({'theta_E': 0.01, 'gamma': 1., 'e1': -0.5,
                          'e2': -0.5, 'center_x': -10, 'center_y': -10})
kwargs_upper_lens.append({'theta_E': 10., 'gamma': 3., 'e1': 0.5, 'e2': 0.5,
                          'center_x': 10, 'center_y': 10})

# add external SHEAR
fixed_lens.append({'ra_0': 0, 'dec_0': 0})
kwargs_lens_init.append({'gamma1': 0.15, 'gamma2': 0.03})
kwargs_lens_sigma.append({'gamma1': 0.05, 'gamma2': 0.05})
kwargs_lower_lens.append({'gamma1': -0.3, 'gamma2': -0.3})
kwargs_upper_lens.append({'gamma1': 0.3, 'gamma2': 0.3})


lens_params = [kwargs_lens_init, kwargs_lens_sigma, 
               fixed_lens, kwargs_lower_lens, kwargs_upper_lens]

# lens galaxy's light model
fixed_lens_light = []
kwargs_lens_light_init = []
kwargs_lens_light_sigma = []
kwargs_lower_lens_light = []
kwargs_upper_lens_light = []


# First Sersic profile
fixed_lens_light.append({'n_sersic': 1.0})
kwargs_lens_light_init.append({'R_sersic': 0.5, 'n_sersic': 1, 'e1': 0.2, 'e2': 0, 
                               'center_x': 0., 'center_y': 0, 'amp': 10})
kwargs_lens_light_sigma.append({'n_sersic': 0, 'R_sersic': 0.3, 'e1': 0.05, 
                                'e2': 0.05, 'center_x': 0.1, 'center_y': 0.1, 
                                'amp': 10})
kwargs_lower_lens_light.append({'e1': -0.5, 'e2': -0.5, 'R_sersic': 0.001, 
                                'n_sersic': .5, 'center_x': -10, 'center_y': -10, 
                                'amp': 0})
kwargs_upper_lens_light.append({'e1': 0.5, 'e2': 0.5, 'R_sersic': 10, 
                                'n_sersic': 1., 
                                'center_x': 10, 'center_y': 10, 'amp': 100})

# Second Sersic profile
fixed_lens_light.append({'n_sersic': 4.0})
kwargs_lens_light_init.append({'R_sersic': 0.5, 'n_sersic': 4, 'e1': 0.2, 'e2': 0, 
                               'center_x': 0., 'center_y': 0, 'amp': 10})
kwargs_lens_light_sigma.append({'n_sersic': 1, 'R_sersic': 0.3, 'e1': 0.05, 
                                'e2': 0.05, 'center_x': 0.1, 'center_y': 0.1, 
                                'amp': 10})
kwargs_lower_lens_light.append({'e1': -0.5, 'e2': -0.5, 'R_sersic': 0.001, 
                                'n_sersic': 4., 'center_x': -10, 'center_y': -10, 
                                'amp': 0})
kwargs_upper_lens_light.append({'e1': 0.5, 'e2': 0.5, 'R_sersic': 10, 
                                'n_sersic': 4., 
                                'center_x': 10, 'center_y': 10, 'amp': 100})

# Join centroids of the two Sersic profiles
joint_lens_with_light = [[0, 0, ['center_x', 'center_y', 'e1', 'e2']], 
                         [0, 1, ['center_x', 'center_y', 'e1', 'e2']]]


lens_light_params = [kwargs_lens_light_init, kwargs_lens_light_sigma, 
                     fixed_lens_light, kwargs_lower_lens_light, 
                     kwargs_upper_lens_light]


from lenstronomy.LightModel.light_model import LightModel
lens_light_model = LightModel(light_model_list=['SERSIC_ELLIPSE', 'SERSIC_ELLIPSE'])



# source galaxy's light model
fixed_source = []
kwargs_source_init = []
kwargs_source_sigma = []
kwargs_lower_source = []
kwargs_upper_source = []


source_model = LightModel(light_model_list=source_model_list)
source_model_list += ['SHAPELETS']


# source galaxy's light model
fixed_source.append({'n_sersic': 1.})
kwargs_source_init = [{'n_sersic': 1, 'R_sersic': 0.23, 'e1': 0, 'e2': 0, 
                       'center_x': 0., 'center_y': 0.}]
kwargs_source_sigma = [{'n_sersic': 0.5, 'R_sersic': 0.1, 'e1': 0.05, 
                        'e2': 0.05, 'center_x': 0.2, 'center_y': 0.2}]
kwargs_lower_source = [{'e1': -0.5, 'e2': -0.5, 'R_sersic': 0.001, 
                        'n_sersic': .5, 'center_x': -10, 'center_y': -10}]
kwargs_upper_source = [{'e1': 0.5, 'e2': 0.5, 'R_sersic': 10, 
                        'n_sersic': 5., 'center_x': 10, 'center_y': 1}]

# SHAPELETS profile
fixed_source.append({'n_max': 5})
kwargs_source_init.append({'beta': 0.29, 'center_x': 0., 'center_y': 0.,
                           'n_max': 5 })
kwargs_source_sigma.append({'beta': 0.01, 'center_x': 0.2, 'center_y': 0.2,
                            'n_max': 5 })
kwargs_lower_source.append({'beta': 0.001, 'center_x': -10, 'center_y': -10,
                            'n_max': 5 })
kwargs_upper_source.append({'beta': 0.1, 'center_x': 10, 'center_y': 10,
                            'n_max': 5 })

joint_source_with_source = [[0, 1, ['center_x', 'center_y']]]

source_params = [kwargs_source_init, 
                 kwargs_source_sigma, 
                 fixed_source, 
                 kwargs_lower_source, 
                 kwargs_upper_source]

# combining all the above specification in the `kwargs_params` dictionary
kwargs_params = {'lens_model': lens_params,
                 'source_model': source_params,
                 'lens_light_model': lens_light_params}

kwargs_constraints = {'joint_lens_with_light': joint_lens_with_light,
                      'joint_source_with_source': joint_source_with_source}
     `

**Comment by mirsazzathossain** [comment]
In, **Join centroids of the two Sersic profiles** you have to add the following:
```python
joint_lens_with_light = [[0, 0, ['center_x', 'center_y', 'e1', 'e2']]]
joint_lens_light_with_lens_light = [[0, 1, ['center_x', 'center_y']]]
```
Insted of:
```python
joint_lens_with_light = [[0, 0, ['center_x', 'center_y', 'e1', 'e2']],
[0, 1, ['center_x', 'center_y', 'e1', 'e2']]]
```

And in `kwargs_model` you have to add `'lens_light_model_list': lens_light_model_list` as a dict item. The `kwargs_model` should be like the following:
```python
kwargs_model = {
    'lens_model_list': lens_model_list,
    'source_light_model_list': source_model_list,
    'lens_light_model_list': lens_light_model_list
}
```

**Comment by Rajit13** [comment]
The best result I got so far with minimum reduced X^2=1.53. the code can be found [here](https://github.com/Rajit13/BDLensing/blob/ad667711278ba70d913b713312656191a8570b4d/DESIJ2158+0257/BDLensing_DESIJ2158+0257_Trial2.ipynb)
What should I try next @ajshajib ?

![1 5](https://user-images.githubusercontent.com/81336262/232464950-bec9fd17-2c0d-40e2-afce-cb9c0407b429.png)

**Reply by ajshajib** [reply, marked as answer]
This is looking great! The arc and the counter image are largely recreated. 👏👏 

To improve some of the residuals, you can try increasing `n_max` and see how much that helps. After that, you can unmask the 7'oclock blob and its counter-image and model them as a second source component.

**Reply by ajshajib** [reply]
By the way, please add `with_caustics=True, scale_size=0.5` in the `model_plot.source_plot()` function call.

**Reply by Rajit13** [reply]
Here is my latest result which produced reduced $\chi^2=1.56$. But I am consistently getting a region with blue pixels around 4 o'clock. I tried increasing the $n_{max}=10$. How to improve that region now any suggestions? 
What are these yellow lines in the reconstructed source plot?

![1 56](https://user-images.githubusercontent.com/81336262/233168056-75f243f4-6427-434c-94ec-31d39bcff60e.png)

Again after seeing the below plot, I don't think the blob opposite 7 o'clock is part of the lens therefore I didn't unmask that. Should I unmask and try again?
![lens light](https://user-images.githubusercontent.com/81336262/233169260-5254e8c9-d567-4457-b402-44a68f1f7ccd.png)
